core/src/main/java/com/linecorp/armeria/server/ArmeriaHttpHeaders.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static com.linecorp.armeria.internal.common.ArmeriaHttpUtil.COOKIE_SEPARATOR;
+import static com.linecorp.armeria.internal.common.ArmeriaHttpUtil.COOKIE_SPLITTER;
+import static com.linecorp.armeria.internal.common.ArmeriaHttpUtil.HTTP_TO_HTTP2_HEADER_DISALLOWED_LIST;
+import static com.linecorp.armeria.internal.common.ArmeriaHttpUtil.convertHeaderValue;
+import static com.linecorp.armeria.internal.common.ArmeriaHttpUtil.maybeWebSocketUpgrade;
+import static com.linecorp.armeria.internal.common.ArmeriaHttpUtil.toLowercaseMap;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.RequestHeadersBuilder;
+import com.linecorp.armeria.internal.common.ArmeriaHttpUtil.CaseInsensitiveMap;
+
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.util.AsciiString;
+import io.netty.util.internal.StringUtil;
+
+/**
+ * Custom implementation of {@link HttpHeaders} that delegates to {@link RequestHeadersBuilder}.
+ */
+public final class ArmeriaHttpHeaders extends HttpHeaders {
+
+    private final RequestHeadersBuilder builder;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param builder the builder to delegate to
+     */
+    public ArmeriaHttpHeaders(RequestHeadersBuilder builder) {
+        this.builder = builder;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param builder the builder to delegate to
+     * @param headers the headers to copy
+     */
+    public ArmeriaHttpHeaders(RequestHeadersBuilder builder, HttpHeaders headers) {
+        this.builder = builder;
+        headers.forEach(e -> add(e.getKey(), e.getValue()));
```

Comment:
This copies the Netty headers to Armeria Headers. Unfortunately, it does not match with the goal of this PR.
> there is no need to copy original netty header.

What do you think of lazily creating or mutating fields when necessary?

```java
private final HttpHeaders nettyHeaders;
@Nullable 
private RequestHeaders armeriaHeaders; 

public ArmeriaHttpHeaders(RequestHeadersBuilder builder, HttpHeaders headers) {
   this.headers = headers;
}

@Override
public String get(String name) {
    if (armeriaHeaders != null) {
        return armeriaHeaders.get(name);
    }
    if (needsValidation(name)) {
        // Create Armeria headers and return;
    }
    // Use Netty Headers as is.
    return headers.get(name);
}

public List<Entry<String, String>> entries() {
    // Create Armeria Headers if null
}
public Iterator<Entry<String, String>> iterator() {
    // Create Armeria Headers if null
}
```

Reply:
Thank you for pointing out!
Let me take a look into it :)

Reply:
I've reviewed your idea and bring some thoughts.

Firstly, I totally agree that current approach doesn't resolve the problem of copying headers 🤦 

And specifically against your advice, `Http1RequestDecoder` has plenty of logic that utilize `RequestHeaders`. So I think we need to convert(create) `HttpHeaders` to `RequestHeaders` most of the time even with 'lazy initializing' strategy. Which means it goes back to copying headers situation again.
I'm not sure if 'lazy initialization' strategy could provide significant value. 🤔 **Correct me if I'm wrong.**

With above idea, I'm afraid we need to copy headers somewhere in our system anyway.

> This copies the Netty headers to Armeria Headers. Unfortunately, it does not match with the goal of this PR.
> 
> > there is no need to copy original netty header.
> 
> What do you think of lazily creating or mutating fields when necessary?
> 
> ```java
> private final HttpHeaders nettyHeaders;
> @Nullable 
> private RequestHeaders armeriaHeaders; 
> 
> public ArmeriaHttpHeaders(RequestHeadersBuilder builder, HttpHeaders headers) {
>    this.headers = headers;
> }
> 
> @Override
> public String get(String name) {
>     if (armeriaHeaders != null) {
>         return armeriaHeaders.get(name);
>     }
>     if (needsValidation(name)) {
>         // Create Armeria headers and return;
>     }
>     // Use Netty Headers as is.
>     return headers.get(name);
> }
> 
> public List<Entry<String, String>> entries() {
>     // Create Armeria Headers if null
> }
> public Iterator<Entry<String, String>> iterator() {
>     // Create Armeria Headers if null
> }
> ```